File: src/main/java/leetcode/longest_substring/LongestSubstring.java
package leetcode.longest_substring;

import java.util.HashSet;
import java.util.Set;

public class LongestSubstring {

    public static void main(String[] args) {
        System.out.println(getLengthOfLongestSubstring("abcabcbb"));
    }

    public static int getLengthOfLongestSubstring(String input) {
        Set<Character> characters = new HashSet<>();
        int length = 0;
        int left = 0;
        int right = 0;
        while (right < input.length()) {
            char c = input.charAt(right);
            while (characters.contains(c)) {
                characters.remove(input.charAt(left));
                left++;
            }
            characters.add(c);
            if (length < characters.size())
                length = characters.size();
            right++;
        }
        return length;
    }
}

File: src/main/java/leetcode/longest_substring/LongestSubstring.kt
package leetcode.longest_substring

fun main() {
    println(getLengthOfLongestSubstring("abcabcbb"))
}

fun getLengthOfLongestSubstring(input: String): Int {
    val characters: MutableSet<Char> = HashSet()
    var length = 0
    var left = 0
    for (right in input.indices) {
        val c = input[right]
        while (characters.contains(c)) {
            characters.remove(input[left])
            left++
        }
        characters.add(c)
        if (length < characters.size) length = characters.size
    }
    return length
}